packages/commonwealth/server/migrations/20230217164706-root-id-refactor.js

```diff
@@ -0,0 +1,134 @@
+        // handle cosmosproposals differently
+        if (rootParts[0] === 'cosmosproposal') {
+          rootParts = ['proposal', rootParts[1]];
+        } else {
+          rootParts = [`proposal/${rootParts[0]}`, rootParts[1]];
+        }
+
+        // otherwise map the root_id to a new thread to backlink proposal comments
+        newThreads.set(c.root_id, {
+          address_id: 1,
+          title: `Thread for ${c.chain}'s ${rootParts[0]} ${rootParts[1]}`,
+          kind: 'link',
+          body: '',
+          url: `https://commonwealth.im/${c.chain}/${rootParts[0]}/${rootParts[1]}`,
+          chain: c.chain,
+          read_only: true,
+          stage: 'discussion',
+          canvas_action: null,
+          canvas_hash: null,
+          canvas_session: null,
+          // set the dates to a year back so they dont spam chain's home page
+          created_at: new Date(
+            new Date().setFullYear(new Date().getFullYear() - 1)
+          ),
+          updated_at: new Date(
+            new Date().setFullYear(new Date().getFullYear() - 1)
+          ),
+        });
+      });
+
+      const inserted = await queryInterface.bulkInsert(
+        'Threads',
+        [...newThreads.values()],
+        { returning: true, transaction: t }
+      );
+
+      // create map of rootId -> thread_id
+      const rootToThreadMap = new Map(
+        inserted.map((thread) => {
+          const urlParts = thread.url.split('/');
+          let rootId =
+            urlParts[urlParts.length - 2] + '_' + urlParts[urlParts.length - 1];
+          if (urlParts[urlParts.length - 2] === 'proposal') {
+            rootId =
+              `cosmos${urlParts[urlParts.length - 2]}` +
+              '_' +
+              urlParts[urlParts.length - 1];
+          }
+          return [rootId, thread.id];
+        })
+      );
+
+      // for each proposal comment, set its root_id to thread_id
+      proposalComments[0].forEach((c) => {
+        // implicitly remove all councilCandidate comments to avoid fkcs
+        if (
+          c.root_id.includes('councilcandidate') ||
+          c.root_id.includes('councilmotion')
+        ) {
+          updateQueries.push(
+            queryInterface.sequelize.query(
+              `UPDATE "Comments" SET "root_id"='1' WHERE "id"=${c.id}`,
+              { transaction: t }
+            )
+          );
+          return;
+        }
+        updateQueries.push(
+          queryInterface.sequelize.query(
+            `UPDATE "Comments" SET "root_id" = ${rootToThreadMap.get(
+              c.root_id
+            )} WHERE "id"=${c.id}`,
+            { transaction: t }
+          )
+        );
+      });
+
+      await Promise.all(updateQueries);
+
+      await queryInterface.renameColumn('Comments', 'root_id', 'thread_id', {
+        transaction: t,
+      });
+    });
+
+    // cant be part of the transaction, or it will cause db to deadlock
+    await queryInterface.changeColumn('Comments', 'thread_id', {
```

Comment:
We may want to make this part of a separate migration, so that worst case (if this fails) we end up with all the logical work succeeding, before we make the schema change.

Reply:
In my understanding if anything fails before it, an error will be thrown and it will not perform the transaction. Because this line will not run in that case, nothing will change. So in my understanding this is ok since it is the last command in the up sequence.